lib/evmone/instructions.hpp

```diff
@@ -91,75 +91,82 @@ inline bool check_memory(ExecutionState& state, const uint256& offset, const uin
 namespace instr
 {
 
-inline StopToken stop(ExecutionState& /*state*/) noexcept
+/// The "core" instruction implementations.
+///
+/// These are minimal EVM instruction implementations which assume:
+/// - the stack requirements (overflow, underflow) have already been checked,
+/// - the "base" gas const has already been charged,
+/// - the `stack` pointer points to the EVM stack top element.
+/// Moreover, these implementations _do not_ inform about new stack height
+/// after execution. The adjustment must be performed by the caller.
+namespace core
 {
-    return {EVMC_SUCCESS};
+inline void noop(intx::uint256* /*stack*/) noexcept {}
+inline constexpr auto pop = noop;
+inline constexpr auto jumpdest = noop;
+
+template <evmc_status_code Status>
+inline StopToken stop_impl() noexcept
+{
+    return {Status};
 }
+inline constexpr auto stop = stop_impl<EVMC_SUCCESS>;
+inline constexpr auto invalid = stop_impl<EVMC_INVALID_INSTRUCTION>;
 
-inline void add(ExecutionState& state) noexcept
+inline void add(intx::uint256* stack) noexcept
 {
-    state.stack.top() += state.stack.pop();
+    stack[-1] = stack[0] + stack[-1];
 }
 
-inline void mul(ExecutionState& state) noexcept
+inline void sub(intx::uint256* stack) noexcept
 {
-    state.stack.top() *= state.stack.pop();
+    stack[-1] = stack[0] - stack[-1];
 }
 
-inline void sub(ExecutionState& state) noexcept
+inline void mul(intx::uint256* stack) noexcept
 {
-    state.stack[1] = state.stack[0] - state.stack[1];
-    state.stack.pop();
+    stack[-1] = stack[0] * stack[-1];
 }
 
-inline void div(ExecutionState& state) noexcept
+inline void div(intx::uint256* stack) noexcept
 {
-    auto& v = state.stack[1];
-    v = v != 0 ? state.stack[0] / v : 0;
-    state.stack.pop();
+    stack[-1] = stack[-1] != 0 ? stack[0] / stack[-1] : 0;
```

Comment:
Maybe better to leave it with `auto& v` for consistency with functions just below.